### PR TITLE
[Feature] Multiple panel users

### PR DESCRIPTION
**Pull request by Sartharon**
### What it does
- Allows multiple users to log on to the panel
- Integrated botlogin.txt user by means of the `paneluser` and `panelpassword` properties
- Basic permission system
- Passwords are sent after running through a one way hash function
- No clear text passwords are saved 
- Management via the panel
- Integrated into the console (only basic functions)

### Other changes
- Channel name is shown on the dashboard
image
- The username and channel name are shown in the profile menu / user menu
image
- `helpers.getModal()` additional override options:
_blockClosing: boolean          // Overrides all methods to close the dialog other than using the buttons
removecancel: boolean          // Removes the "Cancel" button from the dialog_
- `helpers.getConfirmDeleteModal()` introduction of override options:
_deleteText: 'Delete',   // Overrides the text displayed on the "Delete" button
blockClosing: boolean   // Overrides all methods to close the dialog other than using the confirm button_
 
### Considerations
I'm open to change these design choices if needed or requested.
- Web socket tokens are not unique to users. Depending on the assigned permissions, either the `webauth` or `webauthRO` token from the bot properties is uses. Further changes to `WsSharedRWTokenAuthenticationHandler.java` would be necessary to accommodate this.
- Phantombot database schema (key - value) has been used. Where the key is the username and the value is a JSON object
- Users do not have their own panel configuration values, thus are forced to use the ones set by the owner
### Limitations
- Poor permission enforcement due to missing server side rendering of the web page (no PHP or similar)
- Currently only `Full Access` and `Read Only` permission. Additions can be made here, but they require further work and can never really be truly enforced due to the full client side rendering
- No management via commands or twitch chat
- The channel logo is shown logo for every user as their "profile picture"

### Functions limited to the owner
The owner is the user defined in the botlogin.txt
- Bot Setup Page
- Bot OAuth Page
- Create new users
- Edit existing users (other than the user defined in the botlogin.txt)
- Reset a user's passwords
- Deleting a user
- Dashboard configuration (Dark Mode, Event sort order, show/hide twitch player, ...)
- Panel user dashboard
image

### Functions for everyone
- General panel access
- Web socket read token
- Password change option
image

### Additional functions for Full Access users
- Restart bot
- Full panel access (excluding panel user management)

### New user creation / login flow
1. A random (temporary) password is created and shown (bot panel and console)
image
2. New user logs on using the temporary password
3. New user is required to change the password upon first logon
image
4. New user is forced to log out and sign in again with their new password
image

### CODE REVIEW

**Comment by Sartharon** [comment]
> devskim found more than 10 potential problems in the proposed changes. Check the [Files changed](/PhantomBot/PhantomBot/pull/3227/files) tab for more details.

Just the crypto framework which is used [Crypto-JS](https://github.com/brix/crypto-js?utm_source=cdnjs&utm_medium=cdnjs_link&utm_campaign=cdnjs_library)

File: resources/web/panel/js/utils/helpers.js [hunk 1]
```diff
@@ -1491,6 +1511,63 @@ $(function () {
...
+                helpers.signOut();
+            }
+            helpers.currentPanelUserData = res;
+        });
+    };
+
+    helpers.getCookieField = function (name) {
+        let decodedCookie = decodeURIComponent(document.cookie),
+            cookieArray = decodedCookie.split(';');
+        name = name + "="
```
Comment (gmt2001) [inline resources/web/panel/js/utils/helpers.js]: Codacy: Missing semicolon.

File: resources/web/panel/js/utils/helpers.js [hunk 2]
```diff
@@ -1491,6 +1511,63 @@ $(function () {
...
+        for(let i = 0; i < cookieArray.length; i++) {
+            let cookie = cookieArray[i].trim();
+
+            if (cookie.indexOf(name) === 0) {
+                return cookie.substring(name.length, cookie.length);
+            }
+        }
+
+        return undefined;
+    }
```
Comment (gmt2001) [inline resources/web/panel/js/utils/helpers.js]: Codacy: Missing semicolon.

File: resources/web/panel/js/utils/helpers.js [hunk 3]
```diff
@@ -1491,6 +1511,63 @@ $(function () {
...
+
+    helpers.getPanelUserNameFromCookie = function () {
+        let b64 = helpers.getCookieField("panellogin");
+        if (b64 === undefined) {
+            return "Phantombot";
+        }
+        let userpass = atob(b64);
+        let colon = userpass.indexOf(':');
+        return userpass.substring(0, colon);
+    }
```
Comment (gmt2001) [inline resources/web/panel/js/utils/helpers.js]: Codacy: Missing semicolon.

File: resources/web/panel/js/utils/helpers.js [hunk 4]
```diff
@@ -1491,6 +1511,63 @@ $(function () {
...
+
+    helpers.signOut = function () {
+        toastr.info('Signing out...', '', {timeOut: 0});
+        socket.close();
+        window.sessionStorage.removeItem("webauth");
+        document.cookie = 'panellogin=; Path=/;' + (window.location.protocol === 'https:' ? ' Secure' : '') +' Expires=Thu, 01 Jan 1970 00:00:01 GMT;';
+        window.location = window.location.origin + window.location.pathname + 'login/#logoutSuccess=true';
+    }
```
Comment (gmt2001) [inline resources/web/panel/js/utils/helpers.js]: Codacy: Missing semicolon.

File: resources/web/panel/js/utils/helpers.js [hunk 5]
```diff
@@ -1491,6 +1511,63 @@ $(function () {
...
+        }
+        let months = ["January", "February", "March", "April", "May", "June",
+                        "July", "August", "September", "October", "November", "December"
+                        ],
+                hours = (date.getHours().toString().length == 1) ? "0" + date.getHours() : date.getHours(),
+                minutes = (date.getMinutes().toString().length == 1) ? "0" + date.getMinutes() : date.getMinutes(),
+                seconds = (date.getSeconds().toString().length == 1) ? "0" + date.getSeconds() : date.getSeconds();
+
+        return hours +  ":" + minutes + ":" + seconds + " " + date.getDate() + " " + months[date.getMonth()] + " "+ date.getFullYear();
+    }
```
Comment (gmt2001) [inline resources/web/panel/js/utils/helpers.js]: Codacy: Missing semicolon.

File: resources/web/panel/js/utils/helpers.js [hunk 6]
```diff
@@ -1491,6 +1511,63 @@ $(function () {
...
+
+    helpers.getDateStringFromDate = function (date) {
+        if (date === undefined) {
+            return 'Invalid Date';
+        }
+        let months = ["January", "February", "March", "April", "May", "June",
+                        "July", "August", "September", "October", "November", "December"
+                        ],
+                hours = (date.getHours().toString().length == 1) ? "0" + date.getHours() : date.getHours(),
```
Comment (gmt2001) [inline resources/web/panel/js/utils/helpers.js]: Codacy: Use strict equals `===`.

File: resources/web/panel/js/utils/helpers.js [hunk 7]
```diff
@@ -1491,6 +1511,63 @@ $(function () {
...
+                minutes = (date.getMinutes().toString().length == 1) ? "0" + date.getMinutes() : date.getMinutes(),
```
Comment (gmt2001) [inline resources/web/panel/js/utils/helpers.js]: Codacy: Use strict equals `===`.

File: resources/web/panel/js/utils/helpers.js [hunk 8]
```diff
@@ -1491,6 +1511,63 @@ $(function () {
...
+                seconds = (date.getSeconds().toString().length == 1) ? "0" + date.getSeconds() : date.getSeconds();
```
Comment (gmt2001) [inline resources/web/panel/js/utils/helpers.js]: Codacy: Use strict equals `===`.

File: source/tv/phantombot/panel/PanelUser/PanelUser.java [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+import tv.phantombot.PhantomBot;
+import tv.phantombot.panel.WsPanelHandler;
+import tv.phantombot.panel.PanelUser.PanelUserHandler.Permission;
+
+/**
+ * Represents a panel user and should be managed through {@link PanelUserHandler}
+ *
+ * @author Sartharon
+ */
+public class PanelUser {
```
Comment (gmt2001) [inline source/tv/phantombot/panel/PanelUser/PanelUser.java]: Codacy: A class which only has private constructors should be final.

File: source/tv/phantombot/panel/PanelUser/PanelUser.java [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+    private Permission permission;
+    private boolean enabled;
+    private Type userType;
+    private long creationDate;
+    private long lastLogin = -1;
+    private boolean hasSetPassword = false;
+    /**
+     * User types
+     */
+    protected static enum Type {
```
Comment (gmt2001) [inline source/tv/phantombot/panel/PanelUser/PanelUser.java]: Codacy: Unnecessary modifier 'static' on enum 'Type': nested enums are implicitly static.

File: source/tv/phantombot/panel/PanelUser/PanelUserHandler.java [hunk 11]
```diff
@@ -0,0 +1,470 @@
...
+/**
+ * {@link PanelUser Panel user} management for the web panel and websocket
+ *
+ * @author Sartharon
+ */
+public final class PanelUserHandler {
+    /**
+     * Database table suffix for the {@link PanelUser panel users}
+     */
+    protected static final String PANELUSERTABLE = "panelUsers";
```
Comment (gmt2001) [inline source/tv/phantombot/panel/PanelUser/PanelUserHandler.java]: Codacy: Avoid protected fields in a final class.  Change to private or package access.

File: source/tv/phantombot/panel/PanelUser/PanelUserHandler.java [hunk 12]
```diff
@@ -0,0 +1,470 @@
...
+    /**
+     * Messages for use as responses to the web panel
+     */
+    public static enum PanelMessage {
```
Comment (gmt2001) [inline source/tv/phantombot/panel/PanelUser/PanelUserHandler.java]: Codacy: Unnecessary modifier 'static' on enum 'PanelMessage': nested enums are implicitly static.

File: source/tv/phantombot/panel/PanelUser/PanelUserHandler.java [hunk 13]
```diff
@@ -0,0 +1,470 @@
...
+        /**
+         * General Success/Success placeholder
+         */
+        Success("Success",false);
+
+        private String message;
+        private boolean isError;
+        private String JSONkey;
+
+        private PanelMessage(String message, boolean isError) {
```
Comment (gmt2001) [inline source/tv/phantombot/panel/PanelUser/PanelUserHandler.java]: Codacy: Unnecessary modifier 'private' on constructor 'PanelMessage(String, boolean)': enum constructors are implicitly private.

File: source/tv/phantombot/panel/PanelUser/PanelUserHandler.java [hunk 14]
```diff
@@ -0,0 +1,470 @@
...
+         */
+        public String getJSONkey() {
+            return this.JSONkey;
+        }
+    }
+
+     /**
+     * {@link PanelUser Panel user} web panel and websocket permission
+     */
+    public static enum Permission{
```
Comment (gmt2001) [inline source/tv/phantombot/panel/PanelUser/PanelUserHandler.java]: Codacy: Unnecessary modifier 'static' on enum 'Permission': nested enums are implicitly static.

File: source/tv/phantombot/panel/PanelUser/PanelUserHandler.java [hunk 15]
```diff
@@ -0,0 +1,470 @@
...
+        ADMINISTRATOR(1, "Full Access"),
+        /**
+         * Read only access
+         */
+        READ_ONLY(2, "Read Only");
+
+        private int value;
+        private String displayName;
+
+        private Permission(int value, String displayName) {
```
Comment (gmt2001) [inline source/tv/phantombot/panel/PanelUser/PanelUserHandler.java]: Codacy: Unnecessary modifier 'private' on constructor 'Permission(int, String)': enum constructors are implicitly private.

File: source/tv/phantombot/panel/PanelUser/PanelUserHandler.java [hunk 16]
```diff
@@ -0,0 +1,470 @@
...
+        String password = user.generateNewPassword();
+        if (!user.save()) {
+            return PanelMessage.SaveError;
+        }
+        return PanelMessage.Success.setResponse(password);
+    }
+
+    private static List<PanelUser> getAllUsers() {
+        DataStore datastore = PhantomBot.instance().getDataStore();
+        List<PanelUser> users = new ArrayList<PanelUser>();
```
Comment (gmt2001) [inline source/tv/phantombot/panel/PanelUser/PanelUserHandler.java]: Codacy: Explicit type arguments can be replaced by Diamond Operator.

File: source/com/gmt2001/httpwsserver/auth/HttpBasicAuthenticationHandler.java [hunk 17]
```diff
@@ -164,6 +154,6 @@ public boolean isAuthorized(ChannelHandlerContext ctx, FullHttpRequest req) {
 
     @Override
     public boolean isAuthorized(String user, String pass) {
-        return user.equalsIgnoreCase(this.user) && pass.equals(this.pass);
+        return PanelUserHandler.checkLogin(user, pass);
```
Comment (gmt2001) [inline source/com/gmt2001/httpwsserver/auth/HttpBasicAuthenticationHandler.java]: Need to `or` this. Class is used by others for non-standard logins

File: source/com/gmt2001/httpwsserver/auth/HttpBasicAuthenticationHandler.java [hunk 18]
```diff
@@ -138,24 +139,13 @@ public boolean isAuthorized(ChannelHandlerContext ctx, FullHttpRequest req) {
...
             auth = cookies.getOrDefault("panellogin", null);
 
             if (auth != null) {
-                String userpass = new String(Base64.getDecoder().decode(auth));
-                int colon = userpass.indexOf(':');
-
-                if (userpass.substring(0, colon).equalsIgnoreCase(user) && userpass.substring(colon + 1).equals(pass)) {
-                    return true;
-                }
+                return PanelUserHandler.checkLoginB64(auth, req.uri());
```
Comment (gmt2001) [inline source/com/gmt2001/httpwsserver/auth/HttpBasicAuthenticationHandler.java]: Need to `or` this. Class is used by others for non-standard logins

File: source/com/gmt2001/httpwsserver/auth/HttpBasicAuthenticationHandler.java [hunk 19]
```diff
@@ -138,24 +139,13 @@ public boolean isAuthorized(ChannelHandlerContext ctx, FullHttpRequest req) {
...
 
         if (auth != null && auth.startsWith("Basic ")) {
-            auth = auth.substring(6);
-            String userpass = new String(Base64.getDecoder().decode(auth));
-            int colon = userpass.indexOf(':');
-
-            if (userpass.substring(0, colon).equalsIgnoreCase(user) && userpass.substring(colon + 1).equals(pass)) {
-                return true;
-            }
+            return PanelUserHandler.checkLoginB64(auth.substring(6), req.uri());
```
Comment (gmt2001) [inline source/com/gmt2001/httpwsserver/auth/HttpBasicAuthenticationHandler.java]: Need to `or` this. Class is used by others for non-standard logins

**Review by gmt2001 — CHANGES_REQUESTED** [review]

**Review by gmt2001 — APPROVED** [review]

**Review by gmt2001 — APPROVED** [review]

**Pull request merged**